File: dataforge-fx/src/main/java/hep/dataforge/fx/configuration/ConfigFX.java
package hep.dataforge.fx.configuration;

import hep.dataforge.names.Named;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.property.SimpleStringProperty;
import javafx.beans.property.StringProperty;
import javafx.beans.value.ObservableBooleanValue;
import javafx.beans.value.ObservableStringValue;
import javafx.collections.ObservableList;

import java.util.Observable;
import java.util.Optional;

/**
 * Created by darksnake on 01-May-17.
 */
public abstract class ConfigFX extends Observable implements Named {


    protected ObjectProperty<ConfigFXNode> parentProperty = new SimpleObjectProperty<>();
    protected StringProperty name = new SimpleStringProperty();

    protected Optional<ConfigFXNode> getParent() {
        return Optional.ofNullable(parentProperty.get());
    }

    public String getDescription() {
        return descriptionProperty().get();
    }

    protected StringProperty nameProperty() {
        return name;
    }

    public String getName() {
        return nameProperty().get();
    }



    public abstract ObservableStringValue descriptionProperty();
    public abstract ObservableBooleanValue descriptorPresent();
    public abstract ObservableBooleanValue valuePresent();
    public abstract ObservableList<ConfigFX> getChildren();

    public abstract void remove();

    protected void invalidate(){
        setChanged();
        notifyObservers();
    }
}

File: kodex/kodex-gui/src/main/kotlin/hep/dataforge/fx/configuration/ConfigFX.kt
package hep.dataforge.fx.configuration

import hep.dataforge.names.Name
import javafx.beans.binding.StringBinding
import javafx.beans.value.ObservableBooleanValue
import javafx.beans.value.ObservableStringValue

/**
 * A node, containing relative representation of configuration node and description
 * Created by darksnake on 01-May-17.
 */
abstract class ConfigFX(val name: String, val parent: ConfigFXNode? = null) {
    /**
     * remove itself from parent
     */
    abstract fun remove()

    abstract val isEmpty: ObservableBooleanValue
    abstract val descriptionProperty: ObservableStringValue

    val nameProperty = object : StringBinding() {
        override fun computeValue(): String {
            return name
        }
    }


    abstract fun invalidate()

    abstract fun invalidateValue(path: Name)

    abstract fun invalidateNode(path: Name)
}
